src/VisionModel/resnet.jl

```diff
@@ -0,0 +1,87 @@
+using Flux
+
+export ResNet18, ResNet34, ResNet50, ResNet101, ResNet152
+
+basicblock(inplanes, outplanes, downsample = false) = downsample ? 
+  Chain(Conv((3, 3), inplanes => outplanes[1], stride = 2),
+        BatchNorm(outplanes[1], relu),
+        Conv((3, 3), outplanes[1] => outplanes[2], stride = 1, pad = 1),
+        BatchNorm(outplanes[2], relu)) : 
+  Chain(Conv((3, 3), inplanes => outplanes[1], stride = 1),
+        BatchNorm(outplanes[1], relu),
+        Conv((3, 3), outplanes[1] => outplanes[2], stride = 1, pad = 1),
+        BatchNorm(outplanes[2], relu))
+
+bottleneck(inplanes, outplanes, downsample = false) = downsample ?
+  Chain(Conv((1, 1), inplanes => outplanes[1], stride = 2),
+        BatchNorm(outplanes[1], relu),
+        Conv((3, 3), outplanes[1] => outplanes[2], stride = 1, pad = 1),
+        BatchNorm(outplanes[2], relu),
+        Conv((1, 1), outplanes[2] => outplanes[3], stride = 1),
+        BatchNorm(outplanes[3], relu)) :
+  Chain(Conv((1, 1), inplanes => outplanes[1], stride = 1),
+        BatchNorm(outplanes[1], relu),
+        Conv((3, 3), outplanes[1] => outplanes[2], stride = 1, pad = 1),
+        BatchNorm(outplanes[2], relu),
+        Conv((1, 1), outplanes[2] => outplanes[3], stride = 1),
+        BatchNorm(outplanes[3], relu))
+
+projection(inplanes, outplanes, stride) = Chain(Conv((1, 1), inplanes => outplanes, stride=stride),
+                                                               BatchNorm((outplanes), relu))
```

Comment:
The 2nd argument to Skip Connection is supposed to operate on a tuple. I believe the definition should be along the lines of 
`(x1, x2) -> cat(chain_layer(x1), x2)`

Reply:
Ah good catch! I can't believe I missed that all this time...

Just to be clear, it should be something like:
```julia
function projection(...)
    shortcut = Chain(...)

    return (x, y) -> x + shortcut(y)
end
```

Something similar will need to done for `identity` where the padding operation is "`shortcut`"

Reply:
Actually, projection shortcuts have learned parameters, right? I don't think this construction will play friendly with functor. Would the only solution then be to make the projection shortcut a struct?

Reply:
I changed to ``projection(inplanes, outplanes, stride) = (x, m) -> m(x), [Conv((1, 1), inplanes => outplanes, stride=stride), BatchNorm((outplanes), relu), init = x) ``

Reply:
No I don't think that will work. The "connection"/"shortcut" part of SkipConnection is meant to by a binary function that accepts two arguments:
1. the output of the block (i.e. the layers being skipped over)
2. the input to the block

I think the version you have is looking for the input and a model?

Reply:
I am not sure if identity function also needs this. It is already following the rule.

Reply:
It doesn't need it for the "else" case where we just do `+`. But it will need it for the "if" case with the padding. The idea is that `shortcut = array -> # cat operation` and you still return `(x, y) -> x + shortcut(y)`. In the else case you can just `return +` which is already a binary function.

Reply:
Well there is a dimension mismatch error with identity function
```
ResNet: Error During Test at /home/nabanita07/FluxModels.jl/test/VisionModel/resnet.jl:8
  Test threw exception
  Expression: size(m(rand(Float32, 256, 256, 3, 50))) == (1000, 50)
  DimensionMismatch("dimensions must match: a has dims (Base.OneTo(31), Base.OneTo(31), Base.OneTo(64), Base.OneTo(50)), b has dims (Base.OneTo(64), Base.OneTo(64), Base.OneTo(64), Base.OneTo(50)), mismatch at 1")
  Stacktrace:
   [1] promote_shape at ./indices.jl:178 [inlined]
   [2] promote_shape(::Array{Float32,4}, ::Array{Float32,4}) at ./indices.jl:169
   [3] +(::Array{Float32,4}, ::Array{Float32,4}) at ./arraymath.jl:45
   [4] (::SkipConnection)(::Array{Float32,4}) at /home/nabanita07/.julia/packages/Flux/IjMZL/src/layers/basic.jl:259
   [5] applychain(::Tuple{SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,AdaptiveMeanPool{4,2},typeof(flatten),Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}}, ::Array{Float32,4}) at /home/nabanita07/.julia/packages/Flux/IjMZL/src/layers/basic.jl:36 (repeats 4 times)
   [6] (::Chain{Tuple{Conv{2,2,typeof(identity),Array{Float32,4},Array{Float32,1}},BatchNorm{typeof(relu),Array{Float32,1},Array{Float32,1},Float32},MaxPool{2,2},SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,SkipConnection,AdaptiveMeanPool{4,2},typeof(flatten),Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}}})(::Array{Float32,4}) at /home/nabanita07/.julia/packages/Flux/IjMZL/src/layers/basic.jl:38
   [7] macro expansion at /home/nabanita07/FluxModels.jl/test/VisionModel/resnet.jl:8 [inlined]
   [8] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
   [9] top-level scope at /home/nabanita07/FluxModels.jl/test/VisionModel/resnet.jl:6
```

Reply:
That doesn't look like an issue with this change. That looks like precisely the errors we wanted to catch by adding testing. It is entirely possible that our model building logic is not correct.

It's hard for me tell what's going on just from that stack trace. Every shortcut uses `+` to add the two paths together, so that trace could be any of the SkipConnections in the network. Though the number of channels = 64, so I expect this is one of the initial SkipConnections in the network. I would commit what you have so that we can locally debug what's going on. Also, you can debug yourself by adding print statements to check the size/dimensions of arrays as they pass through the network.

Reply:
The resnet models not using identity passes the tests. The problem only arises where identity function is used.

Reply:
Yeah, but the error could be because we are constructing the identity models wrong.

Reply:
I don't think our modelling is wrong. just changing the identity function might do good. I am trying and yes commiting this will do good for now. yes, I am checking those locally.